presto-main/src/main/java/com/facebook/presto/sql/Optimizer.java

```diff
@@ -98,6 +101,9 @@ public Plan validateAndOptimizePlan(PlanNode root, PlanStage stage)
         boolean enableVerboseRuntimeStats = SystemSessionProperties.isVerboseRuntimeStatsEnabled(session);
         if (stage.ordinal() >= OPTIMIZED.ordinal()) {
             for (PlanOptimizer optimizer : planOptimizers) {
```

Comment:
Also curious how responsive each `PlanOptimizer` is. If one `PlanOptimizer` takes most of the time (more than 3 minutes sometime), we still have problem that `PlanOptimizer` runs even after it should be interrupted.

Reply:
Correct, unless we also check interrupt in the optimizer.

As commented above, I think this is a good point, just I am not sure how to throw exception in the running thread, and open to suggestions.

Reply:
> Also curious how responsive each PlanOptimizer is.

Generally speaking, non iterative optimizer is usually fast. Iterative optimizer can be problematic but it has its own timer for it hence not a problem.

Reply:
Does any optimizer have blocking IO calls?

Reply:
> Does any optimizer have blocking IO calls?

You should be able to figure it out from code. In any case, most optimizers in other engines have timeouts because they enumerate all possible plans search for best and that can take too long. Presto optimizers don't do that and they are bounded and simple visits on the logical plan (graph) so that's not an issue here.